Update HttpClientFactory docs (#31629)
dotnet/docs: docs/architecture/microservices/implement-resilient-applications/use-httpclientfactory-to-implement-resilient-http-requests.md

However, the issue isn't really with `HttpClient` per se, but with the [default constructor for HttpClient](/dotnet/api/system.net.http.httpclient.-ctor#system-net-http-httpclient-ctor), because it creates a new concrete instance of <xref:System.Net.Http.HttpMessageHandler>, which is the one that has *sockets exhaustion* and DNS changes issues mentioned above.

> To address the issues mentioned above and to make `HttpClient` instances manageable, .NET Core 2.1 introduced two approaches, one of them being <xref:System.Net.Http.IHttpClientFactory>. It is an interface which can be used to configure and create `HttpClient` instances in an app through Dependency Injection (DI). It also provides extensions for Polly-based middleware to take advantage of delegating handlers in HttpClient.

Reviewer: ```suggestion
To address the issues mentioned above and to make `HttpClient` instances manageable, .NET Core 2.1 introduced two approaches, one of them being <xref:System.Net.Http.IHttpClientFactory>. It's an interface that's used to configure and create `HttpClient` instances in an app through Dependency Injection (DI). It also provides extensions for Polly-based middleware to take advantage of delegating handlers in HttpClient.
```

- Manage the lifetime of <xref:System.Net.Http.HttpMessageHandler> to avoid the mentioned problems/issues that can occur when managing `HttpClient` lifetimes yourself.

> [!TIP]

> > The `HttpClient` instances injected by DI, can be disposed of safely, because the associated `HttpMessageHandler` is managed by the factory. As a matter of fact, injected `HttpClient` instances are *Transient* from a DI perspective, while `HttpMessageHandler` instances can be regarded as *Scoped*.  
> > Note that `HttpMessageHandler` instances have their own DI scopes, **not related** to application scopes (e.g. request scopes).

Reviewer: ```suggestion
> The `HttpClient` instances injected by DI, can be disposed of safely, because the associated `HttpMessageHandler` is managed by the factory. As a matter of fact, injected `HttpClient` instances are *Transient* from a DI perspective, while `HttpMessageHandler` instances can be regarded as *Scoped*. The `HttpMessageHandler` instances have their own DI scopes, **not related** to application (or request) scopes.
```

Author: Request scope was supposed to be an example of application-level DI scopes, but I think I did not elaborate enough on what I meant. I meant ASP.NET incoming request scopes.

(The reason I picked that example is because there's a common misconception that HttpClientFactory would respect incoming request scope, see e.g. https://github.com/dotnet/runtime/issues/36574)

Reviewer: We should probably clarify that then more explicitly. I did initially assume that you were talking about ASP.NET request scopes, but how does that differ from the app scope? Do we have any content on that, would a diagram help?

Author: I think we need to add a section that would explain that at least a bit. I will try to come up with some diagram.

Reviewer: Hmmm, well now seeing this diagram I think that might cause more confusion. 😂 Perhaps, just rephrasing it such that it's clear DI scope differs from request scope, and this sounds more like a concern in ASP.NET, not agnostic to all .NET workloads. Remember, in this docset we focus on agnostic workloads.

Author: >  DI scope differs from request scope

no-no-no, it's not a different _entity_, it's a different _instance_ of a scope

Author: > this sounds more like a concern in ASP.NET

It is not. It is a common confusion that users expect the injected client to have the same scoped dependencies as the scope it was resolved from. But it will not.

ASP.NET is just a most common scenario where people stumble upon this kind-of unpleasantly unexpected behavior.

Author: ![Picture8](https://user-images.githubusercontent.com/3184057/194611129-94543da6-3bec-4888-88d3-04e4472f8c6b.png)
Is it better with less rectangles? @IEvangelist

Reviewer: I'm still not a fan, I typically use excalidraw for diagrams to try to create a more consistent look for docs all up. Do you want me to draw this one?

Reviewer: What do you think @gewarren?

Reviewer 2: I think the diagram is fine as-is.

> [!NOTE]

> > The implementation of `IHttpClientFactory` (`DefaultHttpClientFactory`) is tightly tied to the DI implementation in the `Microsoft.Extensions.DependencyInjection` NuGet package. If you need to use `HttpClient` without DI or with other DI implementations, consider using a `static` or singleton `HttpClient` with `PooledConnectionLifetime` set up, see [HttpClient guidelines for .NET](../../../fundamentals/networking/http/httpclient-guidelines.md).

Reviewer: ```suggestion
> The implementation of `IHttpClientFactory` (`DefaultHttpClientFactory`) is tightly tied to the DI implementation in the `Microsoft.Extensions.DependencyInjection` NuGet package. If you need to use `HttpClient` without DI or with other DI implementations, consider using a `static` or singleton `HttpClient` with `PooledConnectionLifetime` set up. For more information, see [HttpClient guidelines for .NET](../../../fundamentals/networking/http/httpclient-guidelines.md).
```
